File: day4_part1_2020.exs
defmodule Day4 do
  def call do
    input = File.read!("input.txt")
    |> String.split("\n\n", trim: true)

    count_valid_passports(input)
  end

  defp count_valid_passports(input) do
    input
    |> Enum.count(&is_valid_passport/1)
  end

  defp is_valid_passport(passport) do
    required_fields = ~w(byr iyr eyr hgt hcl ecl pid)
    fields = passport
             |> String.split(~r/\s+/)
             |> Enum.map(&String.split(&1, ":"))
             |> Enum.map(fn [key, _value] -> key end)

    Enum.all?(required_fields, &(&1 in fields))
  end
end

File: training_data/reflection-tuning/day4_part1_2020.clj
(ns passport-processing
  (:require [clojure.string :as str]
            [clojure.set :as set]))

(def required-fields #{"byr" "iyr" "eyr" "hgt" "hcl" "ecl" "pid"})

(defn parse-passport [passport-str]
  (->> (str/split passport-str #"\s+")
       (map #(str/split % #":"))
       (into {})))

(defn valid-passport? [passport]
  (set/subset? required-fields (set (keys passport))))

(defn count-valid-passports [input]
  (->> (str/split input #"\n\n")
       (map parse-passport)
       (filter valid-passport?)
       count))

(defn -main []
  (let [input (slurp "input.txt")
        valid-count (count-valid-passports input)]
    (println valid-count)))

(-main)
